cudamapper/src/main.cu

```diff
@@ -105,148 +118,177 @@ int main(int argc, char* argv[])
-        overlapper.get_overlaps(overlaps_to_write.back(), anchors, index_query, index_target);
-        if (0 == overlaps_to_write.back().size())
+
+    struct query_target_range
+    {
+        std::pair<std::int32_t, int32_t> query_range;
+        std::vector<std::pair<std::int32_t, int32_t>> target_ranges;
+    };
+
+    //First generate all the ranges independently, then loop over them.
+    std::vector<query_target_range> query_target_ranges;
+
+    for (std::int32_t query_start_index = 0; query_start_index < queries; query_start_index += index_size)
+    {
+
+        std::int32_t query_end_index = std::min(query_start_index + index_size, static_cast<size_t>(queries));
+
+        query_target_range q;
+        q.query_range = std::make_pair(query_start_index, query_end_index);
+
+        std::int32_t target_start_index = 0;
+        // If all_to_all mode, then we can optimzie by starting the target sequences from the same index as
+        // query because all indices before the current query index are guaranteed to have been processed in
+        // a2a mapping.
+        if (all_to_all)
         {
-            overlaps_to_write.pop_back();
+            target_start_index = query_start_index;
         }
-        overlaps_writer_mtx.unlock();
-    };
 
-    // Start async thread for writing out PAF
-    auto overlaps_writer_func = [&overlaps_to_write, &overlaps_writer_mtx]() {
-        while (true)
+        for (; target_start_index < targets; target_start_index += target_index_size)
         {
-            bool done = false;
-            overlaps_writer_mtx.lock();
-            if (!overlaps_to_write.empty())
-            {
-                CGA_NVTX_RANGE(profile, "overlaps_writer");
-                std::vector<claragenomics::cudamapper::Overlap>& overlaps = overlaps_to_write.front();
-                // An empty overlap vector indicates end of processing.
-                if (overlaps.size() > 0)
-                {
-                    claragenomics::cudamapper::Overlapper::print_paf(overlaps);
-                    overlaps_to_write.pop_front();
-                    overlaps_to_write.shrink_to_fit();
-                }
-                else
-                {
-                    done = true;
-                }
-            }
-            overlaps_writer_mtx.unlock();
-            if (done)
+            std::int32_t target_end_index = std::min(target_start_index + target_index_size,
+                                                     static_cast<size_t>(targets));
+            q.target_ranges.push_back(std::make_pair(target_start_index, target_end_index));
+        }
+
+        query_target_ranges.push_back(q);
+    }
+
+    // This is a per-device cache, if it has the index it will return it, if not it will generate it, store and return it.
+    std::vector<std::map<std::pair<uint64_t, uint64_t>, std::shared_ptr<std::unique_ptr<claragenomics::cudamapper::Index>>>> index_cache(num_devices);
```

Comment:
`std::shared_ptr<std::unique_ptr<...>>` seems redundant, do you need that? Otherwise you can construct a shared pointer directly from an unique pointer via `std::shared_ptr<...> p(std::move(unique_p));`.

Reply:
The reason I am using `std::shared_ptr` is that I was unable to populate the vector with unique_ptrs, (since they don't implement a copy constructor being unique). 

Could I just use `std::unique_ptr<...>` instead by using std::move ?

Reply:
You could, but you'd end up with a similar problem. When you try to use an `Index` instance you could not just say "std::unique_ptr<Index> index = foo_map[bar]" as you'd try to copy an `unique_ptr`. In this case you're trying to share index between several matchers and the caching container, so `shared_ptr` is the right choice. @ahehn-nv is right in saying that you should just move that `unique_ptr` into initial `shared_ptr`

Reply:
OK that makes sense

Reply:
Done

Reply:
I thought you really want shared_ptr semantics (i.e. shared ownership of the object), if there will always be just a single pointer that has the ownership, I would use `std::unique_ptr` and construct the vector with `std::move`.